tools/applets/schedule.py

```diff
@@ -0,0 +1,210 @@
+        self.table.addAction(terminate_pipeline)
+
+        initialization = QtWidgets.QAction("Run initalization experiment", self.table)
+        initialization.triggered.connect(self.run_initialization)
+        self.table.addAction(initialization)
+
+        background = QtWidgets.QAction("Run background experiment", self.table)
+        background.triggered.connect(self.run_background)
+        self.table.addAction(background)
+
+    async def connect_subscribers(self):
+        localhost = "::1"
+        port_notify = 3250
+        self._schedule_sub = ModelSubscriber(
+            "schedule", schedule.Model, self._report_disconnect
+        )
+        await self._schedule_sub.connect(localhost, port_notify)
+
+    def _report_disconnect(self):
+        if not self._disconnected_reported:
+            logging.error("connection to master lost, restart dashboard to reconnect")
+            self.setDisabled(True)
+        self._disconnect_reported = True
+
+    async def labrad_connected(self):
+        await self.artiq_connected()
+        await self.setup_cxn_listeners()
+
+    async def setup_cxn_listeners(self):
+        self.cxn.add_on_connect("artiq", self.run_in_labrad_loop(self.artiq_connected))
+        self.cxn.add_on_disconnect("artiq", self.artiq_disconnected)
+
+    async def artiq_connected(self):
+        self.artiq = self.cxn.get_server("artiq")
+        if not self._disconnect_reported:
+            self.setDisabled(False)
+
+    def artiq_disconnected(self):
+        self.setDisabled(True)
+
+    def set_model(self, model):
+        self.table_model = model
+        self.table.setModel(self.table_model)
+
+    async def delete(self, rid, graceful):
+        """Deletes or terminates an experiment.
+
+        Uses the artiq server functions to delete or to terminate the experiment.
+        The functions also clear the experiment parameters cached in the parameter bank.
+        """
+        if graceful:
+            await self.artiq.request_terminate_experiment(rid)
+        else:
+            await self.artiq.delete_experiment(rid)
+
+    def delete_clicked(self, graceful):
+        idx = self.table.selectedIndexes()
+        if idx:
+            row = idx[0].row()
+            rid = self.table_model.row_to_key[row]
+            if graceful:
+                logger.info("Requested termination of RID %d", rid)
+            else:
+                logger.info("Deleted RID %d", rid)
+            self.run_in_labrad_loop(self.delete)(rid, graceful)
+
+    async def request_term_multiple(self, rids):
+        """Terminates multiple experiments."""
+        for rid in rids:
+            try:
+                await self.artiq.request_terminate_experiment(rid)
+            except Exception as e:
+                # May happen if the experiment has terminated by itself
+                # while we were terminating others.
+                logger.warning(
+                    "failed to request termination of RID %d", rid, exc_info=True
```

Comment:
it may be helpful to add what you have in the comment in the actual error message, ex: "failed to request termination of RID %d, experiment may have terminated itself while in queue"

Reply:
The code is mostly copied from artiq (as said in the docstring of the class). The message here and comments are directly from artiq too. I prefer not to add extra information as there can be multiple reasons that an experiment cannot be terminated (e.g. artiq server is off, artiq master is down, Sinara devices are disconnected).